packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,238 @@
+        )
+      }
+
+      return condition
+    })
+  }
+
+  static resolveConditionType_(data: UpsertDiscountConditionInput):
+    | {
+        type: DiscountConditionType
+        resource_ids: string[]
+      }
+    | undefined {
+    switch (true) {
+      case !!data.products?.length:
+        return {
+          type: DiscountConditionType.PRODUCTS,
+          resource_ids: data.products!,
+        }
+      case !!data.product_collections?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_COLLECTIONS,
+          resource_ids: data.product_collections!,
+        }
+      case !!data.product_types?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TYPES,
+          resource_ids: data.product_types!,
+        }
+      case !!data.product_tags?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TAGS,
+          resource_ids: data.product_tags!,
+        }
+      case !!data.customer_groups?.length:
+        return {
+          type: DiscountConditionType.CUSTOMER_GROUPS,
+          resource_ids: data.customer_groups!,
+        }
+      default:
+        return undefined
+    }
+  }
+
+  async upsertCondition(
+    discountIdOrDiscount: string | Discount,
+    data: UpsertDiscountConditionInput
+  ): Promise<void> {
+    let resolvedConditionType
+
+    return await this.atomicPhase_(
+      async (manager: EntityManager) => {
+        resolvedConditionType =
+          DiscountConditionService.resolveConditionType_(data)
+
+        if (!resolvedConditionType) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Missing one of products, collections, tags, types or customer groups in data`
+          )
+        }
+
+        const discountRepo: DiscountRepository = manager.getCustomRepository(
+          this.discountRepository_
+        )
+
+        let discount
+        // in case id is passed, we retrieve the discount
+        if (typeof discountIdOrDiscount === `string`) {
+          discount = await discountRepo.findOne({
+            where: { id: discountIdOrDiscount },
+          })
+
+          if (!discount) {
+            throw new MedusaError(
+              MedusaError.Types.NOT_FOUND,
+              `Discount with ${discountIdOrDiscount} was not found`
+            )
+          }
+        } else {
+          discount = discountIdOrDiscount
+
+          // ensure discount has rule_id, will also ensure that object is in fact discount
+          if (!discount?.rule_id) {
+            throw new MedusaError(
+              MedusaError.Types.INVALID_DATA,
+              `Discount is missing rule_id`
+            )
+          }
+        }
+
+        const discountConditionRepo: DiscountConditionRepository =
+          manager.getCustomRepository(this.discountConditionRepository_)
+
+        if (data.id) {
+          // throws if condition or discount does not exist
```

Comment:
should not be necessary, the method return `never` as well (if we take in count one of my previous comment) and eventually it should be part of the target method description in the jsdoc?